File: OS4/OS4cpp/legitOS4.cpp
#include "pch.h"
#include <iostream>
using namespace std;

int main(){
	setlocale(LC_ALL, "Russian");

	int i, sizeArr, buff = 0, distinct = 0;
	cout << "Длина массива: ";
	cin >> sizeArr;
	int *arr = new int[sizeArr];

	for (i = 0; i < sizeArr; i++) {
		cin >> arr[i];
	}

	for (int i = 0; i < sizeArr; i++) {
		if (arr[i] != -1) {
			buff = arr[arr[i]];
			arr[arr[i]] = -1;
		}

		while (buff != -1) {
			buff = arr[buff];
			arr[buff] = -1;
		}
	}

	for (int i = 0; i < sizeArr; i++) {
		if (arr[i] == -1)
			distinct++;
	}
	cout << distinct;
	return 0;
}

File: OS4/OS4rb/legitOS4.rb
# encoding: UTF-8
print ("Array lenght: ")
sizeArr =  gets.to_i
buff = 0
unique = 0
arr = Array.new(sizeArr)
for i in 0..(sizeArr-1) do arr[i] = gets.to_i end

for i in 0..(sizeArr-1)
  if (arr[i] != -1) then
    buff  = arr[arr[i]]
    arr[arr[i]] = -1
  end
  until (buff != -1)
    buff = arr[buff]
    arr[buff] = -1
  end end
for i in 0..(sizeArr-1)
  if (arr[i] == -1) then
    unique += 1
  end end
print("Number of distinct elements: ",  unique)
